drivers/pcie/host/msi.c

```diff
@@ -268,18 +270,18 @@ bool pcie_msi_enable(pcie_bdf_t bdf,
 
 	base = pcie_get_cap(bdf, PCI_CAP_ID_MSI);
 
-	if (IS_ENABLED(CONFIG_PCIE_MSI_X)) {
```

Comment:
Do I understand correctly that use of `IS_ENABLED` is forbidden by the rules mentioned in title ?
If this is true, a mail to devel mailinglist would be welcome

Reply:
MISRA Rule 14.3 asks to avoid compile time known condition.
MISRA Rule 2.1 asks (among other things) to avoid unreachable branches (thus very strictly related to the above rule).

In the rare cases where a macro has a conditional on an expression that might be a compile time known condition or not, I've introduced and used a macro called ``CONSTEXPR`` to mean "I'm aware that condition might be constant, it is deliberate".

Where instead the use of a ``#if``, ``COND_CODE_1``, ``IF_ENABLED`` is feasible I've replaced the runtime condition with a preprocessor conditional.

This does not mean that in all cases ``IS_ENABLED`` is forbidden, but only that when code can be improved avoiding to possibly embark unreachable code and/or to put inadvertently constant conditions in code, it is best to take this opportunity and obtain MISRA compliance.

Reply:
So, do I understand correctly that it should not be used on `CONFIG_FOO` anymore ?
In that case, please fix the macro doc which is all about use with CONFIG symbols:
```
/**
 * @brief Check for macro definition in compiler-visible expressions
 *
 * This trick was pioneered in Linux as the config_enabled() macro. It
 * has the effect of taking a macro value that may be defined to "1"
 * or may not be defined at all and turning it into a literal
 * expression that can be handled by the C compiler instead of just
 * the preprocessor. It is often used with a @p CONFIG_FOO macro which
 * may be defined to 1 via Kconfig, or left undefined.
 *
 * That is, it works similarly to <tt>\#if defined(CONFIG_FOO)</tt>
 * except that its expansion is a C expression. Thus, much <tt>\#ifdef</tt>
 * usage can be replaced with equivalents like:
 *
 *     if (IS_ENABLED(CONFIG_FOO)) {
 *             do_something_with_foo
 *     }
 *
 * This is cleaner since the compiler can generate errors and warnings
 * for @p do_something_with_foo even when @p CONFIG_FOO is undefined.
``` 

and then provide a clear example in which is could be used.
Otherwise you'll have to come back on this endlessly.

Reply:
Although I personally think that the benefits to reduce the compiled/possibly embarked code outweigh by far the convenience to have a configured out branch verified for the lack of syntax errors by the compiler (that is a very poor kind of "testing"), I don't think to have the authority to mandate that.
I think that the steering committee should take an official decision about that, before I can change such comment.

Reply:
@nashif what do you think ?